beacon-chain/blockchain/head.go

```diff
@@ -322,36 +324,126 @@ func (s *Service) notifyNewHeadEvent(
 // This saves the attestations inside the beacon block with respect to root `orphanedRoot` back into the
 // attestation pool. It also filters out the attestations that is one epoch older as a
 // defense so invalid attestations don't flow into the attestation pool.
-func (s *Service) saveOrphanedAtts(ctx context.Context, orphanedRoot [32]byte) error {
+func (s *Service) saveOrphanedAtts(ctx context.Context, orphanedRoot, newHeadRoot [32]byte) error {
 	if !features.Get().CorrectlyInsertOrphanedAtts {
 		return nil
 	}
 
-	orphanedBlk, err := s.cfg.BeaconDB.Block(ctx, orphanedRoot)
-	if err != nil {
+	commonAncestorRoot, err := s.commonAncestorRoot(ctx, orphanedRoot, newHeadRoot)
+	switch {
+	// Exit early if there's no common ancestor as there would be nothing to save.
+	case errors.Is(err, errNoCommonAncestorRoot):
+		return nil
+	case err != nil:
 		return err
 	}
 
-	if orphanedBlk == nil || orphanedBlk.IsNil() {
-		return errors.New("orphaned block can't be nil")
+	orphanedBlks, err := s.getOrphanedBlocks(ctx, orphanedRoot, commonAncestorRoot)
+	if err != nil {
+		return err
 	}
 
-	for _, a := range orphanedBlk.Block().Body().Attestations() {
-		// Is the attestation one epoch older.
-		if a.Data.Slot+params.BeaconConfig().SlotsPerEpoch < s.CurrentSlot() {
-			continue
-		}
-		if helpers.IsAggregated(a) {
-			if err := s.cfg.AttPool.SaveAggregatedAttestation(a); err != nil {
-				return err
+	return s.saveAttestations(orphanedBlks)
+}
+
+func (s *Service) saveAttestations(blocks []block.SignedBeaconBlock) error {
+	for _, b := range blocks {
+		for _, a := range b.Block().Body().Attestations() {
+			// Is the attestation one epoch older.
+			if a.Data.Slot+params.BeaconConfig().SlotsPerEpoch < s.CurrentSlot() {
+				continue
 			}
-		} else {
-			if err := s.cfg.AttPool.SaveUnaggregatedAttestation(a); err != nil {
-				return err
+			if helpers.IsAggregated(a) {
+				if err := s.cfg.AttPool.SaveAggregatedAttestation(a); err != nil {
+					return err
+				}
+			} else {
+				if err := s.cfg.AttPool.SaveUnaggregatedAttestation(a); err != nil {
+					return err
+				}
 			}
+			saveOrphanedAttCount.Inc()
 		}
-		saveOrphanedAttCount.Inc()
 	}
 
 	return nil
 }
+
+// Get the blocks from the head to the base.
+// It includes the head block but excludes the base block
+func (s *Service) getOrphanedBlocks(ctx context.Context, head, base [32]byte) ([]block.SignedBeaconBlock, error) {
```

Comment:
In golang, we prefer idiomatic naming and that generally means not using "get" prefix for getter methods. This should be simply `orphanedBlocks(...)`

```suggestion
func (s *Service) orphanedBlocks(ctx context.Context, head, base [32]byte) ([]block.SignedBeaconBlock, error) {
```